bevy_ecs::event::EventId should manually implement Eq, PartialEq, Ord, PartialOrd, and Hash

Currently, `bevy_ecs::event::EventId` makes use of derive macros for its implementations of a set of traits:

```rust
#[derive(Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct EventId<E: Event> {
    pub id: usize,
    _marker: PhantomData<E>,
}
```

This creates an issue when attempting to actually *use* these implementations (especially that of `Hash`), as it requires them to be implemented on the events themselves; despite the generic only being used in a marker type.

It would be better to manually implement these traits to ignore the marker, freeing them up to be used in more places than they are currently:

```rust
error[E0599]: the method `insert` exists for struct `Local<'_, HashSet<EventId<MouseButtonInput>>>`, but its trait bounds were not satisfied
   --> src\interaction.rs:327:54
    |
327 |  if !old_events.contains(&id) && processed_events.insert(id) {
    |                                                   ^^^^^^
    |
   ::: D:\Programs\Rust\cargo\registry\src\github.com-1ecc6299db9ec823\bevy_ecs-0.10.0\src\event.rs:20:1
    |
20  | pub struct EventId<E: Event> {
    | ---------------------------- doesn't satisfy `bevy::ecs::event::EventId<MouseButtonInput>: Hash`
    |
    = note: the following trait bounds were not satisfied:
            `bevy::ecs::event::EventId<MouseButtonInput>: Hash`
```

In the above error message, I am unable to insert an `EventId` object into a `HashSet` as `MouseButtonInput` does not implement `Hash`. As `MouseButtonInput` is only being used in a marker, I don't believe it should be relied upon in the `Hash` implementation; however the derive macro is not smart enough to know this. For this reason, I believe the traits should be implemented manually.

--- a/crates/bevy_ecs/src/event.rs
+++ b/crates/bevy_ecs/src/event.rs
@@ -4,7 +4,14 @@ use crate as bevy_ecs;
 use crate::system::{Local, Res, ResMut, Resource, SystemParam};
 use bevy_utils::detailed_trace;
 use std::ops::{Deref, DerefMut};
-use std::{fmt, hash::Hash, iter::Chain, marker::PhantomData, slice::Iter};
+use std::{
+    cmp::Ordering,
+    fmt,
+    hash::{Hash, Hasher},
+    iter::Chain,
+    marker::PhantomData,
+    slice::Iter,
+};
 /// A type that can be stored in an [`Events<E>`] resource
 /// You can conveniently access events using the [`EventReader`] and [`EventWriter`] system parameter.
 ///
@@ -16,7 +23,6 @@ impl<T> Event for T where T: Send + Sync + 'static {}
 ///
 /// An `EventId` can among other things be used to trace the flow of an event from the point it was
 /// sent to the point it was processed.
-#[derive(Eq, PartialEq, Ord, PartialOrd, Hash)]
 pub struct EventId<E: Event> {
     pub id: usize,
     _marker: PhantomData<E>,
@@ -43,6 +49,32 @@ impl<E: Event> fmt::Debug for EventId<E> {
             std::any::type_name::<E>().split("::").last().unwrap(),
             self.id,
         )
+    }
+}
+
+impl<E: Event> PartialEq for EventId<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl<E: Event> Eq for EventId<E> {}
+
+impl<E: Event> PartialOrd for EventId<E> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<E: Event> Ord for EventId<E> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+impl<E: Event> Hash for EventId<E> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Hash::hash(&self.id, state);
     }
 }
 

